src/model.cpp

```diff
@@ -472,4 +577,208 @@ ModelPtr Model::clone() const
+        importedComponentCopy->setName(component->name());
+
+        // Temporarily add component to new model to find units used.
+        auto tempModel = Model::create();
+        tempModel->addComponent(importedComponentCopy);
+        tempModel->linkUnits();
+        std::vector<UnitsPtr> requiredUnits;
+        for (size_t i = 0; i < tempModel->unitsCount(); ++i) {
+            auto u = tempModel->units(i);
+            requiredUnits.push_back(u);
+        }
+
+        // Make a map of component name to component pointer.
+        ComponentNameMap newComponentNames = createComponentNamesMap(importedComponentCopy);
+        for (const auto &entry : newComponentNames) {
+            std::string newName = entry.first;
+            size_t count = 1;
+            while (std::find(componentNames.begin(), componentNames.end(), newName) != componentNames.end()) {
+                newName += "_" + convertToString(count++);
+            }
+            if (newName != entry.first) {
+                entry.second->setName(newName);
+            }
+        }
+
+        // If the component 'component' has variables then they are equivalent variables and they
+        // need to be exchanged with the real variables from the component 'importedComponent'.
+        for (size_t i = 0; i < component->variableCount(); ++i) {
+            auto placeholderVariable = component->variable(i);
+            for (size_t j = 0; j < placeholderVariable->equivalentVariableCount(); ++j) {
+                auto localModelVariable = placeholderVariable->equivalentVariable(j);
+                auto importedComponentVariable = importedComponentCopy->variable(placeholderVariable->name());
+                Variable::removeEquivalence(placeholderVariable, localModelVariable);
+                Variable::addEquivalence(importedComponentVariable, localModelVariable);
+            }
+        }
+        parent->replaceComponent(index, importedComponentCopy);
+
+        // Apply the rebased equivalence map onto the modified model.
+        applyEquivalenceMapToModel(rebasedMap, model);
+
+        // Copy over units used in imported component to this model.
+        for (const auto &u : requiredUnits) {
+            if (!model->hasUnits(u)) {
+                size_t count = 1;
+                while (!model->hasUnits(u) && model->hasUnits(u->name())) {
+                    auto name = u->name();
+                    name += "_" + convertToString(count++);
+                    u->setName(name);
+                }
+                model->addUnits(u);
+            }
+        }
+    }
+}
+
+void flattenComponentTree(const ComponentEntityPtr &parent, const ComponentPtr &component, size_t componentIndex)
+{
+    flattenComponent(parent, component, componentIndex);
+
+    for (size_t index = 0; index < component->componentCount(); ++index) {
+        auto c = component->component(index);
+        flattenComponentTree(component, c, index);
+    }
+}
+
+void Model::flatten()
+{
```

Comment:
Again for consistency ... should this be flattenModel() (like validateModel and parseModel ?)

Reply:
Hmm, indeed. This means that we also need to rename `Model::clone()` to `Model::cloneModel()`. Done.

Reply:
this is a method on `Model`, so I don't think it needs to have the model, that is by definition what is being flattened or cloned.

Reply:
Oops, I agree with @nickerso. For the `Validator` and `Parser` classes, it's a different story since we are indeed validating/parsin a model, not a validator/parser.